### PR TITLE
Asset-swapper 0.0.1 merge

### PR DESCRIPTION
**Pull request by dave4506**
## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-buyer/src/swap_quoter.ts [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+    public readonly orderRefreshIntervalMs: number;
+    public readonly expiryBufferMs: number;
+    private readonly _contractWrappers: ContractWrappers;
+    // cache of orders along with the time last updated keyed by assetData
+    private readonly _ordersEntryMap: ObjectMap<OrdersEntry> = {};
+    /**
+     * Instantiates a new SwapQuoter instance given existing liquidity in the form of orders and feeOrders.
+     * @param   supportedProvider       The Provider instance you would like to use for interacting with the Ethereum network.
+     * @param   orders                  A non-empty array of objects that conform to SignedOrder. All orders must have the same makerAssetData and takerAssetData.
+     * @param   feeOrders               A array of objects that conform to SignedOrder. All orders must have the same makerAssetData (ZRX) and takerAssetData. Defaults to an empty array.
```
Comment (dekz) [inline packages/asset-buyer/src/swap_quoter.ts]: appears to be missing in params

File: packages/asset-buyer/src/swap_quoter.ts [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+     *
+     * @return  An object that conforms to SwapQuote that satisfies the request. See type definition for more information.
+     */
+    public async getMarketSellSwapQuoteAsync(
+        makerAssetData: string,
+        takerAssetData: string,
+        takerAssetSellAmount: BigNumber,
+        options: Partial<SwapQuoteRequestOpts> = {},
+    ): Promise<MarketSellSwapQuote> {
+        assert.isBigNumber('makerAssetBuyAmount', takerAssetSellAmount);
```
Comment (dekz) [inline packages/asset-buyer/src/swap_quoter.ts]:
```suggestion
        assert.isBigNumber('takerAssetSellAmount', takerAssetSellAmount);
```

File: packages/asset-buyer/src/swap_quoter.ts [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+        this._contractWrappers = new ContractWrappers(this.provider, {
+            networkId,
+        });
+    }
+    /**
+     * Get a `SwapQuote` containing all information relevant to fulfilling a swap between a desired ERC20 token address and ERC20 owned by a provided address.
+     * You can then pass the `SwapQuote` to a `SwapQuoteConsumer` to execute a buy, or process SwapQuote for on-chain consumption.
+     * @param   makerAssetData           The makerAssetData of the desired asset to swap for (for more info: https://github.com/0xProject/0x-protocol-specification/blob/master/v2/v2-specification.md).
+     * @param   takerAssetData           The takerAssetData of the asset to swap makerAssetData for (for more info: https://github.com/0xProject/0x-protocol-specification/blob/master/v2/v2-specification.md).
+     * @param   takerAssetSellAmount     The amount of maker asset to swap for.
```
Comment (dekz) [inline packages/asset-buyer/src/swap_quoter.ts]:
```suggestion
     * @param   takerAssetSellAmount     The amount of taker asset to sell. 
```

File: packages/asset-buyer/src/swap_quoter.ts [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+            options,
+        )) as MarketSellSwapQuote;
+    }
+
+    /**
+     * Get a `SwapQuote` containing all information relevant to fulfilling a swap between a desired ERC20 token address and ERC20 owned by a provided address.
+     * You can then pass the `SwapQuote` to a `SwapQuoteConsumer` to execute a buy, or process SwapQuote for on-chain consumption.
+     * @param   makerAssetData           The makerAssetData of the desired asset to swap for (for more info: https://github.com/0xProject/0x-protocol-specification/blob/master/v2/v2-specification.md).
+     * @param   takerAssetData           The takerAssetData of the asset to swap makerAssetData for (for more info: https://github.com/0xProject/0x-protocol-specification/blob/master/v2/v2-specification.md).
+     * @param   makerAssetSwapAmount     The amount of maker asset to swap for.
```
Comment (dekz) [inline packages/asset-buyer/src/swap_quoter.ts]:
```suggestion
     * @param   makerAssetBuyAmount     The amount of maker asset to swap for.
```

File: packages/asset-buyer/src/swap_quoter.ts [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+     * @param   options                  Options for the request. See type definition for more information.
+     *
+     * @return  An object that conforms to SwapQuote that satisfies the request. See type definition for more information.
+     */
+    public async getMarketBuySwapQuoteForERC20TokenAddressAsync(
```
Comment (dekz) [inline packages/asset-buyer/src/swap_quoter.ts]: do you think these are simpler functions to use over the developer encoding and using `getMarketBuySwapQuote`? might be worth considering the inverse of descriptive function names so this one is easier to use

File: packages/asset-buyer/src/swap_quoter.ts [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+        const swapQuote = this.getMarketBuySwapQuoteAsync(makerAssetData, takerAssetData, makerAssetBuyAmount, options);
+        return swapQuote;
+    }
+
+    /**
+     * Get a `SwapQuote` containing all information relevant to fulfilling a swap between a desired ERC20 token address and ERC20 owned by a provided address.
+     * You can then pass the `SwapQuote` to a `SwapQuoteConsumer` to execute a buy, or process SwapQuote for on-chain consumption.
+     * @param   makerAssetData           The makerAssetData of the desired asset to swap for (for more info: https://github.com/0xProject/0x-protocol-specification/blob/master/v2/v2-specification.md).
+     * @param   takerAssetData           The takerAssetData of the asset to swap makerAssetData for (for more info: https://github.com/0xProject/0x-protocol-specification/blob/master/v2/v2-specification.md).
+     * @param   makerAssetSwapAmount     The amount of maker asset to swap for.
```
Comment (dekz) [inline packages/asset-buyer/src/swap_quoter.ts]:
```suggestion
     * @param   takerAssetSellAmount     The amount of taker asset to sell.
```

File: packages/asset-buyer/src/swap_quoter.ts [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+     * @param   options             Options for the request. See type definition for more information.
+     *
+     * @return  An object that conforms to LiquidityForAssetData that satisfies the request. See type definition for more information.
+     */
+    public async getLiquidityForMakerTakerAssetDataPairAsync(
+        makerAssetData: string,
+        takerAssetData: string,
+        options: Partial<LiquidityRequestOpts> = {},
+    ): Promise<LiquidityForAssetData> {
+        const shouldForceOrderRefresh =
```
Comment (dekz) [inline packages/asset-buyer/src/swap_quoter.ts]: merge the global default here

File: packages/asset-buyer/src/swap_quoter.ts [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+    /**
+     * Grab orders from the map, if there is a miss or it is time to refresh, fetch and process the orders
+     * @param   makerAssetData      The makerAssetData of the desired asset to swap for (for more info: https://github.com/0xProject/0x-protocol-specification/blob/master/v2/v2-specification.md).
+     * @param   takerAssetData      The takerAssetData of the asset to swap makerAssetData for (for more info: https://github.com/0xProject/0x-protocol-specification/blob/master/v2/v2-specification.md).
+     * @param shouldForceOrderRefresh  If set to true, new orders and state will be fetched instead of waiting for the next orderRefreshIntervalMs.
+     */
+    public async getOrdersAndFillableAmountsAsync(
+        makerAssetData: string,
+        takerAssetData: string,
+        shouldForceOrderRefresh: boolean,
```
Comment (dekz) [inline packages/asset-buyer/src/swap_quoter.ts]: default the param to the global default?

File: packages/asset-buyer/src/swap_quoter.ts [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+        assert.isNumber('slippagePercentage', slippagePercentage);
+        const zrxTokenAssetData = this._getZrxTokenAssetDataOrThrow();
+        const isMakerAssetZrxToken = makerAssetData === zrxTokenAssetData;
+        // get the relevant orders for the makerAsset and fees
+        // if the requested assetData is ZRX, don't get the fee info
+        const [ordersAndFillableAmounts, feeOrdersAndFillableAmounts] = await Promise.all([
+            this.getOrdersAndFillableAmountsAsync(makerAssetData, takerAssetData, shouldForceOrderRefresh),
+            isMakerAssetZrxToken
+                ? Promise.resolve(constants.EMPTY_ORDERS_AND_FILLABLE_AMOUNTS)
+                : this.getOrdersAndFillableAmountsAsync(zrxTokenAssetData, takerAssetData, shouldForceOrderRefresh),
```
Comment (dekz) [inline packages/asset-buyer/src/swap_quoter.ts]: is this only a requirement for the forwarder?

Comment (dave4506) [reply packages/asset-buyer/src/swap_quoter.ts]: this is part of how fees are taken from the taker if the makerAsset happens to be denominated in the feeToken (in v2's case that would only be the case for ZRX). Later as v3 goes in affect (hopefully), the logic that we conditionally use for `isMakerAssetZrxToken` will probably become the defacto logic for asset-swapper for all asset pairs.

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+                        shouldValidate: true,
+                    },
+                );
+            }
+            return txHash;
+        } catch (err) {
+            if (_.includes(err.message, ContractWrappersError.SignatureRequestDenied)) {
+                throw new Error(SwapQuoteConsumerError.SignatureRequestDenied);
+            } else if (_.includes(err.message, ForwarderWrapperError.CompleteFillFailed)) {
+                throw new Error(SwapQuoteConsumerError.TransactionValueTooLow);
```
[the remark this answers is not in the document]
Comment (dave4506) [reply packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: While, I didn't write this particular code, my understanding is that we are catching only a few errors we think are particularly hard for developers to understand.

**Pull request merged**